ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/DensityDisconnect.hs

```diff
@@ -0,0 +1,280 @@
+
+    peer forkPrefix forkSuffix chain =
+      EvolvingPeer {
+        forkPrefix = toHeaders forkPrefix,
+        forkSuffix = toHeaders forkSuffix,
+        candidate = AF.Empty AF.AnchorGenesis,
+        suffix = AF.toOldestFirst headers,
+        tip = branchTip chain,
+        prefixSlots = lastSlot forkPrefix,
+        killed = False
+      }
+      where
+        headers = toHeaders chain
+
+    trunk = btTrunk gtBlockTree
+
+    branches = btBranches gtBlockTree
+
+data MonotonicityResult =
+  HonestKilled
+  |
+  Nonmonotonic
+  |
+  Finished
+  deriving Show
+
+-- | Check whether the honest peer was killed or a peer's new losing state violates monotonicity, i.e. if it was found
+-- to be losing before, it shouldn't be found winning later.
+--
+-- If that is the case, return @Left (False, peers)@ to indicate that the test is over and failed.
+--
+-- Otherwise, remove all adversaries that either have no more blocks or have more than @sgen@ slots after their fork
+-- intersection.
+--
+-- If no adversaries remain, return @Left (True, peers)@ to indicate that the test is over and succeeded.
+--
+-- Otherwise, return @Right remaining@ to continue with the next step.
+updatePeers ::
+  GenesisWindow ->
+  PeerId ->
+  [PeerId] ->
+  Peers EvolvingPeer ->
+  Either (MonotonicityResult, Peers EvolvingPeer) (Peers EvolvingPeer)
+updatePeers (GenesisWindow sgen) target disconnect peers
+  | HonestPeer `elem` disconnect
+  = Left (HonestKilled, peers)
+  | killed peer
+  , not (target `elem` disconnect)
+  = Left (Nonmonotonic, peers)
+  | null remaining
+  = Left (Finished, peers)
+  | otherwise
+  = Right peers {others = remaining}
+  where
+    Peer {value = peer} = getPeer target peers
+
+    remaining = Map.filter (not . discardPeer) (others peers)
+
+    discardPeer Peer {value = EvolvingPeer {candidate, suffix, prefixSlots}} =
+      null suffix || lastSlot candidate - prefixSlots > fromIntegral sgen
+
+-- | Find the earliest intersection, used to compute the selection.
+firstBranch :: Peers EvolvingPeer -> Peer EvolvingPeer
+firstBranch peers =
+  minimumBy (compare `on` predicate) (toList (others peers))
+  where
+    predicate = fromWithOrigin 0 . AF.anchorToSlotNo . AF.anchor . forkSuffix . value
+
+-- | Take one block off the peer's suffix and append it to the candidate fragment.
+--
+-- Since we don't remove the honest peer when it's exhausted, this may be called with an empty suffix.
+movePeer :: EvolvingPeer -> EvolvingPeer
+movePeer = \case
+  peer@EvolvingPeer {candidate, suffix = h : t} -> peer {candidate = candidate AF.:> h, suffix = t}
+  peer -> peer
+
+-- | Repeatedly run the GDD, each time updating a random peer to advance by one block.
+-- The selection is set to the first k blocks of the first fork.
+-- The tips are the last blocks of each full branch.
+-- The returned 'Bool' indicates whether the honest peer won and no monotonicity violations were detected.
+evolveBranches ::
```

Comment:
Is this roughly what you had in mind, @amesgen?

Reply:
Yeah, exactly!

FTR:
> One idea for a GDD test: Test that the GDD logic is "consistent under chain extension":
> For a given set of candidate chains (plus "claims more headers" information), the GDD logic will for every peer either decide to disconnect or to keep them for now ("inconclusive"). Now consider a specific peer. As more and more information about its candidate chain gets available for the GDD logic (ie we receive more extending headers), the GDD might eventually change its judgement from "inconclusive" to "disconnect", but never the other way around. A property test could make sure that the "disconnect"->"inconclusive" judgement state transition is never occurring.